### PR TITLE
Future/peggy2create bridge token

### PR DESCRIPTION
**Pull request by daniportugalgit**
Before committing, I merged this branch with `future/peggy2` and `future/peggy2chain-id`, so now we have a single PR with all changes from `future/peggy2create-bridge-token` and `future/peggy2chain-id`.

Tasks included:
1) [Allow CosmosBridge to Create a New Bridge Token in submitProphecyClaimAggregatedSigs](https://app.zenhub.com/workspaces/sifchain-603ed608675da3001744d736/issues/sifchain/sifnode/1668)

2) [Reimplement Import Whitelist#1646](https://app.zenhub.com/workspaces/sifchain-603ed608675da3001744d736/issues/sifchain/sifnode/1646)

3) [Batch prophecy claim: add network descriptor#1664](https://app.zenhub.com/workspaces/sifchain-603ed608675da3001744d736/issues/sifchain/sifnode/1664)

I suggest abandoning [PR #1663](https://github.com/Sifchain/sifnode/pull/1663) in favor of this PR, which already has adjustments according to #1663 PR code review.

### CODE REVIEW

File: smart-contracts/contracts/CosmosBridge.sol [hunk 1]
```diff
@@ -242,8 +255,18 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
...
 
-        // if we are double pegging, then we are going to need to get the address on this chain
-        address tokenAddress = claimData.doublePeg ? sourceAddressToDestinationAddress[claimData.tokenAddress] : claimData.tokenAddress;
+        // if we are double pegging AND we don't control the token, we deploy a new smart contract
+        address tokenAddress;
+        if(claimData.doublePeg && !_isManagedToken(claimData.tokenAddress)) {
+            string memory name;
+            string memory symbol;
+            uint8 decimals;
+            (name, symbol, decimals) = BridgeBank(bridgeBank).getTokenData(claimData.tokenAddress);
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridge.sol:264]: If the token doesn't exist on this chain, how are we going to be able to get this data from the bridgebank? This doesn't make sense.

File: smart-contracts/contracts/CosmosBridge.sol [hunk 2]
```diff
@@ -259,20 +282,44 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
...
      */
     function createNewBridgeToken(
         string calldata symbol,
         string calldata name,
         address sourceChainTokenAddress,
         uint8 decimals,
-        uint256 chainDescriptor
-    ) external onlyValidator {
+        uint256 _networkDescriptor
+    ) external onlyValidator returns(address) {
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridge.sol:306]: We should remove this function as we only want to be creating new tokens through this function: `_submitProphecyClaimAggregatedSigs`

File: smart-contracts/contracts/CosmosBridgeStorage.sol [hunk 3]
```diff
@@ -58,8 +59,13 @@ contract CosmosBridgeStorage {
         uint256 amount;
     }
 
+    /**
+    * @notice network descriptor
+    */
+    uint256 public networkDescriptor;
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridgeStorage.sol:65]: Again, let's test an upgrade off a mainnnet fork with the proper oz config file loaded up and ensure that the framework plays nice with this change.

File: smart-contracts/test/testBridgeBank.js [hunk 4]
```diff
@@ -87,24 +91,32 @@ describe("Test Bridge Bank", function () {
 
     it("should not allow user to lock ERC20 tokens", async function () {
       const FakeToken = await ethers.getContractFactory("FakeERC20");
-      const fakeToken = await FakeToken.deploy();
+      fakeToken = await FakeToken.deploy();
+      
+      // Add the token into white list
+      await state.bridgeBank.connect(operator)
+      .updateEthWhiteList(fakeToken.address, true)
```
Comment (ElliotFriedman) [inline smart-contracts/test/testBridgeBank.js:98]: let's indent here

File: smart-contracts/test/testBridgeBank.js [hunk 5]
```diff
@@ -87,24 +91,32 @@ describe("Test Bridge Bank", function () {
...
 
       // Approve and lock tokens
       await expect(state.bridgeBank.connect(userOne).lock(state.sender, fakeToken.address, state.amount))
         .to.emit(state.bridgeBank, 'LogLock')
-        .withArgs(userOne.address, state.sender, fakeToken.address, state.amount, "3", 18, "", "");
+        .withArgs(userOne.address, state.sender, fakeToken.address, state.amount, "3", 18, "", "", state.networkDescriptor);
 
       /*
-                  msg.sender,
+            msg.sender,
```
Comment (ElliotFriedman) [inline smart-contracts/test/testBridgeBank.js:107]: Remove commented code.

File: smart-contracts/test/testCosmosBridge.js [hunk 6]
```diff
@@ -306,5 +314,165 @@ describe("Test Cosmos Bridge", function () {
...
+            claimData,
+            signatures
+        )).to.emit(state.cosmosBridge, 'LogNewBridgeTokenCreated')
+        .withArgs(18, state.networkDescriptor, state.name, state.symbol, state.token.address, expectedAddress);
+
+      const newlyCreatedTokenAddress = await state.cosmosBridge.sourceAddressToDestinationAddress(state.token.address);
+      expect(newlyCreatedTokenAddress).to.be.equal(expectedAddress);
+
+      /* @SOL
+        emit LogNewBridgeTokenCreated(
```
Comment (ElliotFriedman) [inline smart-contracts/test/testCosmosBridge.js:362]: Let's delete if not needed.

File: smart-contracts/test/testCosmosBridge.js [hunk 7]
```diff
@@ -306,5 +314,165 @@ describe("Test Cosmos Bridge", function () {
...
+        nonce: state.nonce,
+        networkDescriptor: state.networkDescriptor
+      };
+
+      await state.cosmosBridge
+        .connect(userOne)
+        .submitProphecyClaimAggregatedSigs(digest, claimData, signatures);
+
+      const newlyCreatedTokenAddress = await state.cosmosBridge.sourceAddressToDestinationAddress(state.token.address);
+      expect(newlyCreatedTokenAddress).to.be.equal('0x0000000000000000000000000000000000000000');
```
Comment (ElliotFriedman) [inline smart-contracts/test/testCosmosBridge.js:403]: 0x0000000000000000000000000000000000000000 should be a constant so it is not copied and pasted around the file a bunch.

File: smart-contracts/test/testCosmosBridge.js [hunk 8]
```diff
@@ -63,7 +66,8 @@ describe("Test Cosmos Bridge", function () {
       owner,
       userOne,
       userThree,
-      pauser
+      pauser,
+      networkDescriptor
```
Comment (ElliotFriedman) [inline smart-contracts/test/testCosmosBridge.js:70]: Somewhere in this file we should write a test that queries the networkDescriptor from the cosmos bridge and asserts that it was set correctly.

File: smart-contracts/test/test_networkDescriptorMismatch.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        networkDescriptor: state.networkDescriptor
+      };
+
+      await expect(state.cosmosBridge
+        .connect(userOne)
+        .submitProphecyClaimAggregatedSigs(
+            digest,
+            claimData,
+            signatures
+        )).to.be.revertedWith("INV_NET_DESC");
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_networkDescriptorMismatch.js:132]: Please merge these tests with the tests in the test_security file.

File: smart-contracts/test/test_security.js [hunk 10]
```diff
@@ -70,10 +73,31 @@ describe("Security Test", function () {
...
         userOne,
         userThree,
-        pauser.address
+        pauser.address,
+        networkDescriptor
       );
     });
 
+    it("should allow operator to call reinitalize after initialization", async function () {
+      await expect(state.bridgeBank.connect(operator).reinitialize(
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_security.js:82]: We should add another test that checks if you try to call reinitialize a second time it fails and reverts as you should only be able to call this function a single time.

File: smart-contracts/test/test_security.js [hunk 11]
```diff
@@ -70,10 +73,31 @@ describe("Security Test", function () {
...
+        operator.address,
+        state.cosmosBridge.address,
+        owner.address,
+        pauser.address,
+        state.networkDescriptor
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_security.js:87]: Here let's also assert that the state changed to the expected variables with some expect statements.

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: smart-contracts/contracts/CosmosBridge.sol [hunk 12]
```diff
@@ -136,13 +141,13 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
...
-        for (uint256 i = 0; i < validators.length; i++) {
-            totalPower += getValidatorPower(validators[i].signer);
+    function getSignedPower(SignatureData[] calldata _validators) public view returns(uint256) {
+        uint256 totalPower_ = 0;
+        for (uint256 i = 0; i < _validators.length; i++) {
+            totalPower_ += getValidatorPower(_validators[i].signer);
         }
 
-        return totalPower;
+        return totalPower_;
```
[the remark this answers is not in the document]
Comment (daniportugalgit) [reply smart-contracts/contracts/CosmosBridge.sol:150]: I missed that one. I'll change it to `_totalPower` to avoid variable shadowing (`totalPower` lives in ValsetStorage).

File: smart-contracts/contracts/CosmosBridge.sol (hunk 1 above)
Comment (daniportugalgit) [reply smart-contracts/contracts/CosmosBridge.sol:264]: You're absolutely right. Things are starting to make more sense in my mind. I'll rethink this flow.

File: smart-contracts/contracts/BridgeBank/BridgeBank.sol [hunk 13]
```diff
@@ -76,6 +130,30 @@ contract BridgeBank is BankStorage,
...
+    function updateEthWhiteList(address _token, bool _inList)
+        public
+        onlyOperator
+        returns (bool)
+    {
+        // Do not allow a token with the same address to be whitelisted
+        if (_inList) {
+            // if we want to add it to the whitelist, make sure it's not there yet
+            require(!getTokenInEthWhiteList(_token), "whitelisted");
+            require(!getCosmosTokenInWhiteList(_token), "whitelisted");
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/BridgeBank/BridgeBank.sol:149]: This error message should throw an error like so:
`already in cosmos whitelist`
so that it denotes the whitelist it is a part of.

File: smart-contracts/contracts/CosmosBridge.sol [hunk 14]
```diff
@@ -91,7 +94,11 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
         address tokenAddress,
         uint256 amount,
         bool doublePeg,
-        uint128 nonce
+        uint128 nonce,
+        uint256 _networkDescriptor,
+        string memory tokenName,
+        string memory tokenSymbol,
+        uint8 tokenDecimals
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridge.sol:101]: Do we really want to base our prophecyID off of the tokenName, tokenSymbol and tokenDecimals?

File: smart-contracts/contracts/CosmosBridge.sol [hunk 15]
```diff
@@ -183,6 +198,7 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
         _submitProphecyClaimAggregatedSigs(hashDigest, claimData, signatureData);
     }
 
+    // Essentially, each tx to the submitProphecyClaimAggregatedSigs needs to include a networkDescriptor or other unique identifier that only the smart contract on that chain has. If the chain identifier in the smart contract and the chain identifier in the message do not match, the transaction should fail.
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridge.sol:201]: Please break this comment up into multiple lines for readability.

File: smart-contracts/contracts/CosmosBridge.sol [hunk 16]
```diff
@@ -242,8 +267,20 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
...
+        address tokenAddress;
+        if(claimData.doublePeg && !_isManagedToken(claimData.tokenAddress)) {
+            tokenAddress = _createNewBridgeToken(
+                claimData.tokenSymbol,
+                claimData.tokenName,
+                claimData.tokenAddress,
+                claimData.tokenDecimals,
+                claimData.networkDescriptor
+            );
+        } else {
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridge.sol:280]: We could just make this an else if clause and make sure that this is a double peg tx, then we wouldn't need to do the ternary operator for assigning the tokenAddress.

File: smart-contracts/test/helpers/testFixture.js [hunk 17]
```diff
@@ -200,6 +239,11 @@ async function singleSetup(
     // Approve tokens to contract
     await state.token.connect(userOne).approve(state.bridgeBank.address, state.amount).should.be.fulfilled;
       
+    // Add the token into white list
+    await state.bridgeBank.connect(operator)
+    .updateEthWhiteList(state.token.address, true)
```
Comment (ElliotFriedman) [inline smart-contracts/test/helpers/testFixture.js:244]: Please tab this in.

File: smart-contracts/test/testCosmosBridge.js [hunk 18]
```diff
@@ -306,5 +284,124 @@ describe("Test Cosmos Bridge", function () {
...
+        .connect(userOne)
+        .submitProphecyClaimAggregatedSigs(
+            digest,
+            claimData,
+            signatures
+        )).to.emit(state.cosmosBridge, 'LogNewBridgeTokenCreated')
+        .withArgs(18, state.networkDescriptor, state.name, state.symbol, state.token.address, expectedAddress);
+
+      const newlyCreatedTokenAddress = await state.cosmosBridge.sourceAddressToDestinationAddress(state.token.address);
+      expect(newlyCreatedTokenAddress).to.be.equal(expectedAddress);
```
Comment (ElliotFriedman) [inline smart-contracts/test/testCosmosBridge.js:319]: Nice test!

File: smart-contracts/test/testCosmosBridge.js [hunk 19]
```diff
@@ -306,5 +284,124 @@ describe("Test Cosmos Bridge", function () {
...
+        tokenDecimals: state.decimals,
+        validators: [userOne, userTwo, userFour],
+      });
+
+      await state.cosmosBridge
+        .connect(userOne)
+        .submitProphecyClaimAggregatedSigs(digest, claimData, signatures);
+
+      const newlyCreatedTokenAddress = await state.cosmosBridge.sourceAddressToDestinationAddress(state.token.address);
+      expect(newlyCreatedTokenAddress).to.be.equal(state.constants.zeroAddress);
```
Comment (ElliotFriedman) [inline smart-contracts/test/testCosmosBridge.js:346]: We should also assert that the recipient's balance of the token went up by the amount we specified in the claim.

File: smart-contracts/test/testCosmosBridge.js [hunk 20]
```diff
@@ -306,5 +284,124 @@ describe("Test Cosmos Bridge", function () {
...
+        isDoublePeg: true,
+        nonce: state.nonce + 1,
+        networkDescriptor: state.networkDescriptor,
+        tokenName: state.name,
+        tokenSymbol: state.symbol,
+        tokenDecimals: state.decimals,
+        validators: [userOne, userTwo, userFour],
+      });
+
+      await expect(state.cosmosBridge
```
Comment (ElliotFriedman) [inline smart-contracts/test/testCosmosBridge.js:398]: Earlier in this test, you say that this time it shouldn't emit an event, but here, you are testing that an event was emitted. I'm a bit confused.

File: smart-contracts/test/test_security.js [hunk 21]
```diff
@@ -388,6 +528,11 @@ describe("Security Test", function () {
     });
 
     it("should allow users to unpeg troll token, but then does not receive", async function () {
+      // Add the token into white list
+      await state.bridgeBank.connect(operator)
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_security.js:532]: Let's add another test that updates the eth whitelist with a cosmos whitelisted token and then check to make sure that the error message is different than an already whitelisted ethereum asset.

File: smart-contracts/test/test_txSignatureAggregation.js [hunk 22]
```diff
@@ -186,13 +176,13 @@ describe("Gas Cost Tests", function () {
  * 
  * 
 Unlock Gas Cost With 4 Validators
-tx0  182434
+tx0  165769
 ~~~~~~~~~~~~
-Total:  182434
+Total:  165769
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_txSignatureAggregation.js:181]: Are these numbers up to date? I'm wondering if they changed around after the double pegging logic was added.

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: smart-contracts/contracts/BridgeBank/BankStorage.sol [hunk 23]
```diff
@@ -10,9 +11,9 @@ contract BankStorage is
     CosmosWhiteListStorage {
 
     /*
-    * @notice [DEPRECATED] operator address that can update the smart contract
+    * @notice operator address that can update the smart contract
```
Comment (banshee) [inline smart-contracts/contracts/BridgeBank/BankStorage.sol:14]: Can we make this comment clearer?  Let's say exactly what the operator can do.

File: smart-contracts/contracts/CosmosBridge.sol (hunk 14 above)
Comment (juniuszhou) [reply smart-contracts/contracts/CosmosBridge.sol:101]: The requirement for prophecyID just to make it unique. the minimum implementation is just including the cosmosSender and cosmosSenderSequence. other data items are optional

Comment (juniuszhou) [reply smart-contracts/contracts/CosmosBridge.sol:101]: one question, for not double peg case, if we need tokenName, tokenSymbol and tokenDecimals data in the claimData every time.

Comment (daniportugalgit) [reply smart-contracts/contracts/CosmosBridge.sol:101]: @ElliotFriedman , we don't need them there, I removed them from the prophecyID.
@juniuszhou unfortunately, we do need to pass tokenName, tokenSymbol and tokenDecimals every time. There are alternatives, but they would demand a bigger refactory of the code - we would need to separate those fields from the basic claimData and use an overloaded function to deal with the variation. I'm not sure it's worth it, but discussing it in one of our meetings might be good.

File: smart-contracts/test/test_txSignatureAggregation.js (hunk 22 above)
Comment (daniportugalgit) [reply smart-contracts/test/test_txSignatureAggregation.js]: Yes, they were up to date before I added `tokenName`, `tokenSymbol`, and `tokenDecimals` to `claimData`.
The new values are 173990 gas for lock and 179737 gas for mint (I updated the comments again).

File: smart-contracts/contracts/BridgeBank/CosmosBankStorage.sol [hunk 24]
```diff
@@ -33,8 +34,13 @@ contract CosmosBankStorage {
     */
     mapping(string => string) private lowerToUpperTokens;
 
+    /**
+    * @notice network descriptor
+    */
+    uint256 public networkDescriptor;
```
[the remark this answers is not in the document]
Comment (daniportugalgit) [reply smart-contracts/contracts/BridgeBank/CosmosBankStorage.sol:40]: Ok! I adjusted the code according to the comments and all of them are solved. This is the only one left.
Now I'm going to focus on this upgrade test.

File: smart-contracts/contracts/BridgeBank/BridgeBank.sol [hunk 25]
```diff
@@ -76,6 +130,30 @@ contract BridgeBank is BankStorage,
...
+    function updateEthWhiteList(address _token, bool _inList)
+        public
+        onlyOperator
+        returns (bool)
+    {
+        // Do not allow a token with the same address to be whitelisted
+        if (_inList) {
+            // if we want to add it to the whitelist, make sure it's not there yet
+            require(!getTokenInEthWhiteList(_token), "already in eth whitelist");
+            require(!getCosmosTokenInWhiteList(_token), "already in cosmos whitelist");
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/BridgeBank/BridgeBank.sol:149]: Let's change the cosmos and eth whitelist to have different error messages when a token isn't in the whitelist when trying to lock or burn.

File: smart-contracts/test/test_bridgeBankLock.js [hunk 26]
```diff
@@ -147,6 +175,19 @@ describe("Test Bridge Bank", function () {
...
+      // Add the tokens into white list
+      await state.bridgeBank.connect(operator)
+        .updateEthWhiteList(state.token1.address, true)
+        .should.be.fulfilled;
+
+      await state.bridgeBank.connect(operator)
+        .updateEthWhiteList(state.token2.address, true)
+        .should.be.fulfilled;
+
+      await state.bridgeBank.connect(operator)
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_bridgeBankLock.js:187]: Seeing how you have to send 3 tx's here in order to whitelist 3 separate tokens seems wasteful. This should be a separate PR but we need to have a bulkWhitelist function that handles eth assets to whitelist. This will save us gas on bulk updates.

File: smart-contracts/test/test_bridgeBankLock.js [hunk 27]
```diff
@@ -409,6 +485,19 @@ describe("Test Bridge Bank", function () {
...
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_bridgeBankLock.js:497]: Once we add the bulkWhitelist function, we can update these tests to only hit that one function.

File: smart-contracts/test/test_bridgeBankLock.js [hunk 28]
```diff
@@ -433,4 +522,142 @@ describe("Test Bridge Bank", function () {
...
+        state.bridgeBank.address,
+        state.amount
+      );
+
+      // Attempt to lock tokens
+      await expect(state.bridgeBank.connect(userOne).lock(
+        state.sender,
+        state.token1.address,
+        state.amount
+      )).to.be.revertedWith("Only token in whitelist can be transferred to cosmos");
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_bridgeBankLock.js:585]: We need to change what we expect this to be reverted with as we update the error message to specify which whitelist it should have been a part of.

File: smart-contracts/test/test_security.js [hunk 29]
```diff
@@ -70,10 +74,59 @@ describe("Security Test", function () {
...
+        accounts[5].address, // was owner.address
+        accounts[6].address, // was pauser.address
+        state.networkDescriptor + 1 // was state.networkDescriptor
+      )).to.be.fulfilled;
+
+      expect(await state.bridgeBank.operator()).to.equal(accounts[3].address);
+      expect(await state.bridgeBank.cosmosBridge()).to.equal(accounts[4].address);
+      expect(await state.bridgeBank.owner()).to.equal(accounts[5].address);
+      expect(await state.bridgeBank.pausers(accounts[6].address)).to.equal(true);
+      expect(await state.bridgeBank.networkDescriptor()).to.equal(state.networkDescriptor + 1);
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_security.js:96]: Good looks

File: smart-contracts/test/test_txSignatureAggregation.js [hunk 30]
```diff
@@ -60,9 +62,15 @@ describe("Gas Cost Tests", function () {
       owner,
       userOne,
       userThree,
-      pauser
+      pauser,
+      networkDescriptor
     );
 
+    // Add the token into white list
+    await state.bridgeBank.connect(operator)
```
Comment (ElliotFriedman) [inline smart-contracts/test/test_txSignatureAggregation.js:70]: I see that pretty often we have to add these tokens to the eth whitelist. We may consider making a test fixture that wraps another test fixture and just add the tokens to the eth whitelist.

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: smart-contracts/contracts/CosmosBridge.sol [hunk 31]
```diff
@@ -231,19 +256,27 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
         }
 
         uint256 signedPower = getSignedPower(signatureData);
-
         require(getProphecyStatus(signedPower), "INV_POW");
+        //require(getProphecyStatus(getSignedPower(signatureData)), "INV_POW");
```
Comment (ElliotFriedman) [inline smart-contracts/contracts/CosmosBridge.sol:260]: Remove commented code.

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

**Review by ElliotFriedman — APPROVED** [review]
LGTM!

Next steps for this new code will be adding new gas profiling and seeing if there are any ways we can cut down on gas costs here.
